### PR TITLE
Checking real parameters

### PR DESCRIPTION
**Pull request by petko-boyadzhiev**
Checking service parameters via a REST call to see if the service needs to be updated.

### CODE REVIEW

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/DetermineServiceCreateUpdateServiceActionsStep.java [hunk 1]
```diff
@@ -43,36 +44,32 @@
...
-public class DetermineServiceCreateUpdateServiceActionsStep extends SyncFlowableStep {
+@Component("determineServiceCreateUpdateActionsStep") public class DetermineServiceCreateUpdateServiceActionsStep extends SyncFlowableStep {
 
     private static final String LAST_SERVICE_OPERATION = "last_operation";
     private static final String SERVICE_OPERATION_TYPE = "type";
     private static final String SERVICE_OPERATION_STATE = "state";
 
-    @Inject
-    private ServiceGetter serviceInstanceGetter;
+    @Inject private ServiceGetter serviceInstanceGetter;
```
Comment (enchobelezirev) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/DetermineServiceCreateUpdateServiceActionsStep.java]: Is it from the IntelliJ or the new formatter?

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/DetermineServiceCreateUpdateServiceActionsStep.java [hunk 2]
```diff
@@ -244,12 +238,23 @@ private boolean shouldUpdateTags(CloudServiceExtended service, List<String> exis
...
         }
-        existingServiceTags = ObjectUtils.defaultIfNull(existingServiceTags, Collections.<String> emptyList());
+        existingServiceTags = ObjectUtils.defaultIfNull(existingServiceTags, Collections.<String>emptyList());
         return !existingServiceTags.equals(service.getTags());
     }
 
-    private boolean shouldUpdateCredentials(CloudServiceExtended service, Map<String, Object> credentials) {
-        return !Objects.equals(service.getCredentials(), credentials);
+    private boolean shouldUpdateCredentials(CloudServiceExtended service, CloudService existingService, CloudControllerClient client) {
+        if (existingService == null) {
```
Comment (enchobelezirev) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/DetermineServiceCreateUpdateServiceActionsStep.java]: The existing service will be never null here. Check the if statement where this method is invoked.

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/DetermineServiceCreateUpdateServiceActionsStep.java [hunk 3]
```diff
@@ -244,12 +238,23 @@ private boolean shouldUpdateTags(CloudServiceExtended service, List<String> exis
...
+            return MapUtils.isEmpty(service.getCredentials());
+        }
+        try {
+            Map<String, Object> serviceParameters = client.getServiceParameters(existingService.getMetadata().getGuid());
+            getStepLogger().debug("Existing service parameters: " + secureSerializer.toJson(serviceParameters));
+            return !Objects.equals(service.getCredentials(), serviceParameters);
```
Comment (enchobelezirev) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/DetermineServiceCreateUpdateServiceActionsStep.java]: Why are we checking for equality between credentials and parameters?
Shouldn't we make the corresponding changes to our cloud model?

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/DetermineServiceCreateUpdateServiceActionsStep.java [hunk 4]
```diff
@@ -244,12 +238,23 @@ private boolean shouldUpdateTags(CloudServiceExtended service, List<String> exis
...
+        } catch (RestClientException e) {
+            getStepLogger().warn(e, Messages.CANNOT_RETRIEVE_SERVICE_PARAMETERS, service.getName());
+            //TODO: Optimization that should be deprecated at some point. So here is a todo for that.
```
Comment (enchobelezirev) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/DetermineServiceCreateUpdateServiceActionsStep.java]: This is not an optimization - this is a hack

File: pom.xml [hunk 5]
```diff
@@ -1,1076 +1,1076 @@
...
+        <spring.security.oauth2.version>2.3.5.RELEASE</spring.security.oauth2.version>
+        <spring.security.jwt.version>1.0.9.RELEASE</spring.security.jwt.version>
+        <spring.cloud.version>1.2.3.RELEASE</spring.cloud.version>
+        <hana.cloud.version>1.0.4.RELEASE</hana.cloud.version>
+        <quartz.version>2.2.1</quartz.version>
+        <jackson.version>2.9.8</jackson.version>
+        <jackson-asl.version>1.9.13</jackson-asl.version>
+        <liquibase.version>3.4.1</liquibase.version>
+        <liquibase-slf4j.version>1.2.1</liquibase-slf4j.version>
+        <cloudfoundry.client.version>1.13.0-SNAPSHOT</cloudfoundry.client.version>
```
Comment (enchobelezirev) [inline pom.xml]: When the PR is merged in the cf-java-client, release it and adopt the new version here, directly :)

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/message/Messages.java [hunk 6]
```diff
@@ -1,489 +1,490 @@
...
+    public static final String ERROR_PREPARING_TO_EXECUTE_TASKS_ON_APP = "Error preparing to execute tasks on application \"{0}\"";
+    public static final String ERROR_EXECUTING_TASK_ON_APP = "Execution of task \"{0}\" failed. Download the application logs \"{1}\", via the dmol command, and check them for more information.";
+    public static final String ERROR_DETECTING_COMPONENTS_TO_UNDEPLOY = "Error detecting components to undeploy";
+    public static final String ERROR_DELETING_IDLE_ROUTES = "Error deleting idle routes";
+    public static final String ERROR_REGISTERING_SERVICE_URLS = "Error registering service URLs";
+    public static final String ERROR_UNREGISTERING_SERVICE_URLS = "Error unregistering service URLs";
+    public static final String ERROR_CREATING_SERVICE_BROKERS = "Error creating service brokers";
+    public static final String ERROR_DELETING_SERVICE_BROKERS = "Error deleting service brokers";
+    public static final String ERROR_UPDATING_SUBSCRIBERS = "Error updating subscribers";
+    public static final String ERROR_NO_SERVICE_URL_SPECIFIED = "No service URL is spe
```
Comment (enchobelezirev) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/message/Messages.java]: Both messages are inconsistent.
Change the second to be : "Cannot retrieve service instance of service with name(optional - decide to keep it or not) \"{0}\""

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/DetermineServiceCreateUpdateServiceActionsStep.java (hunk 1 above)
Comment (petko-boyadzhiev) [reply com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/DetermineServiceCreateUpdateServiceActionsStep.java]: yes

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/DetermineServiceCreateUpdateServiceActionsStep.java (hunk 3 above)
Comment (petko-boyadzhiev) [reply com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/DetermineServiceCreateUpdateServiceActionsStep.java]: will be done in a separate pull req.

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/DetermineServiceCreateUpdateServiceActionsStep.java (hunk 1 above)
Comment (enchobelezirev) [reply com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/DetermineServiceCreateUpdateServiceActionsStep.java]: #totallynotunderstandablecode

Comment (petko-boyadzhiev) [reply com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/DetermineServiceCreateUpdateServiceActionsStep.java]: [https://sagittarius.agency/blog/lame-the-overuse-of-hashtags-in-a-non-twitter-environment](https://sagittarius.agency/blog/lame-the-overuse-of-hashtags-in-a-non-twitter-environment)

**Pull request merged**